src/jit/gentree.cpp

```diff
@@ -13920,8 +13920,8 @@ GenTreePtr Compiler::gtNewTempAssign(unsigned tmp, GenTreePtr val)
     var_types valTyp = val->TypeGet();
     if (val->OperGet() == GT_LCL_VAR && lvaTable[val->gtLclVar.gtLclNum].lvNormalizeOnLoad())
     {
-        valTyp = lvaGetRealType(val->gtLclVar.gtLclNum);
-        val    = gtNewLclvNode(val->gtLclVar.gtLclNum, valTyp, val->gtLclVar.gtLclILoffs);
+        valTyp      = lvaGetRealType(val->gtLclVar.gtLclNum);
+        val->gtType = valTyp;
```

Comment:
It seems that it is implicitly part of the contract of this method that the unmodified `val` will be the rhs of the new assignment. Since `LIR::Use::ReplaceWithLclVar` seems to now have a dependency on this, you should state it explicitly in this method.
You should also add a corresponding assert to `LIR::Use::ReplaceWithLclVar`.